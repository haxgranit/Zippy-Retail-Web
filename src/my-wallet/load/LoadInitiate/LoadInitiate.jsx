/* eslint-disable react/jsx-props-no-spreading */
import { useMsal } from '@azure/msal-react';
// import Select, { components } from 'react-select';
import { useEffect, useState } from 'react';
import { Button, FormControl, Modal } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import Api from '../../../api';
import PageContainer from '../../../common/PageContainer';
import Options from '../options/options';

export default function LoadInitiate() {
  const [fundingSources, setFundingSources] = useState([]);
  const [amount, setAmoumt] = useState();
  const [cardExpired, setCardExpired] = useState(false);
  const [selectedAccount, setSelectedAccount] = useState(undefined);
  const navigate = useNavigate();

  const { instance, accounts } = useMsal();

  useEffect(() => {
    new Api(instance, accounts[0])
      .listFundingSources()
      .then((result) => {
        const options = result.map((s) => ({
          value: s.id,
          ...s,
          label: s.displayName
            ?? (s?.bankAccount?.accountNumber ?? s?.paymentCard?.number),
        }));
        setFundingSources(options ?? []);
        // setSelectedAccount(options[options.findIndex((f) => f.isDefault)]);
      })
      .catch((error) => console.log('error', error));
  }, []);

  const [loadTried, setLoadTried] = useState(false);
  function loadFunds() {
    setLoadTried(true);

    if (!selectedAccount) {
      return;
    }
    if (!(amount > 0)) {
      return;
    }

    if (selectedAccount?.paymentCard) {
      setCardExpired(true);
      return;
    }
    const fundRequest = {
      amount,
      sourceId: selectedAccount,
    };
    new Api(instance, accounts[0])
      .postFundLoadRequest(fundRequest)
      .then(() => {
        navigate('/my-wallet/load/status', {
          state: { ...fundRequest, status: fundRequest.amount > 500 ? 'Success' : 'In_Progress' },
        });
      })
      .catch((error) => {
        console.log('error', error);
        navigate('/my-wallet/load/status', {
          state: { ...fundRequest, status: 'Failure' },
        });
      });
  }

  // const CustomOption = (props) => {
  //   const { data, isSelected } = props;

  //   if (data.bankAccount) {
  //     return (
  //       <components.Option {...props}>
  //         <>
  //           <div className="d-flex align-item-center justify-content-between bg-white">
  //             <div className="bg-white text-black">
  //               <div className="d-flex align-item-center justify-content-between">
  //                 Bank Name: ABC Bank
  //               </div>
  //               <div className="d-flex align-item-center justify-content-between">
  //                 Account Number:
  //                 {' '}
  //                 {data.bankAccount.accountNumber}
  //               </div>
  //             </div>
  //             <input className="" type="radio" defaultChecked={isSelected} />
  //           </div>
  //         </>
  //       </components.Option>
  //     );
  //   }
  //   return (
  //     <components.Option {...props}>
  //       <div className="border-top">
  //         <div className="d-flex align-item-center justify-content-between">
  //           <p>{data.paymentCard.number}</p>
  //           {data.paymentCard.number}
  //           <img width={30} src="https://pngimg.com/uploads/visa/visa_PNG24.png" alt="" />
  //           <input type="radio" defaultChecked={isSelected} />
  //         </div>
  //         <div className="d-flex align-item-center justify-content-between">
  //           Debit Card
  //         </div>
  //       </div>
  //     </components.Option>
  //   );
  // };

  return (
    <PageContainer title="Personal Banking" subTitle="Made Fun With Zippy!">
      <div className="title">
        Load wallet
      </div>

      <FormControl
        className="amount"
        placeholder="0.00"
        type="number"
        step=".01"
        defaultValue={amount}
        onChange={(event) => { setAmoumt(event.target.value); }}
      />
      {loadTried && !(amount > 0)
        && (
          <span style={{ color: 'red' }}>
            {' '}
            *Amount Must Be Greater Than Zero
          </span>
        )}
      {/* <Select
        options={fundingSources}
        components={{
          Option: CustomOption,
        }}
        onChange={(option) => {
          setSelectedAccount(option);
          const seletedAccount = fundingSources.find(
            (source) => source.id === +option.value,
          );
          if (seletedAccount.paymentCard) {
            // eslint-disable-next-line no-alert
            alert('Card choosen, hence validated');
          }
        }}
        defaultValue={selectedAccount}
      /> */}
      <Options
        defaultTitle="Select funding source"
        fundingSources={fundingSources}
        onChange={(option) => {
          setSelectedAccount(option);
          const seletedAccount = fundingSources.find(
            (source) => source.id === +option.value,
          );
          if (seletedAccount.paymentCard) {
          // eslint-disable-next-line no-alert
            alert('Card choosen, hence validated');
          }
        }}
      />
      {loadTried && !selectedAccount
        && (
          <span style={{ color: 'red' }}>
            {' '}
            *Please Select funding source
          </span>
        )}
      <div className="action">
        <Button
          className="zippy-btn"
          onClick={() => {
            loadFunds();
          }}
        >
          Load
        </Button>
      </div>
      <Modal show={cardExpired} onHide={() => setCardExpired(false)}>
        <Modal.Header closeButton>
          <Modal.Title>Error</Modal.Title>
        </Modal.Header>
        <Modal.Body>This Card is expired!</Modal.Body>
        <Modal.Footer>
          <Button
            className="zippy-btn"
            variant="secondary"
            onClick={() => setCardExpired(false)}
          >
            Remove Card
          </Button>
        </Modal.Footer>
      </Modal>
    </PageContainer>
  );
}
