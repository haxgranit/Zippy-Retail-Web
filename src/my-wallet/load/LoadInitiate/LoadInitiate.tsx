import { useMsal } from '@azure/msal-react';
import { useEffect, useState } from 'react';
import { Button, Modal } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import Api, { FundingSource, FundingSourceTransactionResponce } from '../../../api';
import AmountInput from '../../../common/AmountInput';
import PageContainer from '../../../common/PageContainer';
import Options from '../options/options';
import Alert from '../../../common/Alert';

export default function LoadInitiate() {
  const [fundingSources, setFundingSources] = useState<FundingSource[]>([]);
  const [amount, setAmount] = useState(0.00);
  const [cardExpired, setCardExpired] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [selectedAccount, setSelectedAccount] = useState<FundingSource | undefined>(undefined);
  const navigate = useNavigate();

  const { instance, accounts } = useMsal();

  useEffect(() => {
    new Api(instance, accounts[0])
      .listFundingSources()
      .then((result:FundingSource[]) => {
        setFundingSources(result);
      })
      .catch((error) => console.log('error', error));
  }, []);

  const [loadTried, setLoadTried] = useState(false);
  function loadFunds() {
    setLoadTried(true);

    if (!selectedAccount) {
      return;
    }
    if (!(amount > 0)) {
      return;
    }
    const fundRequest = {
      amount,
      sourceId: selectedAccount.id,
    };
    setIsLoading(true);
    new Api(instance, accounts[0])
      .postFundingSourceTransaction(fundRequest)
      .then((resp:FundingSourceTransactionResponce) => {
        setIsLoading(false);
        navigate('/my-wallet/load/status', {
          state: { ...fundRequest, status: resp.status, id: resp.id },
        });
      })
      .catch((error) => {
        console.log('error', error);
        setIsLoading(false);
        navigate('/my-wallet/load/status', {
          state: { ...fundRequest, status: 'Failure' },
        });
      });
  }

  return (
    <PageContainer title="Personal Account" subTitle="Made Fun With Zippy!" backdropImage="backdrop-image-2">
      <AmountInput
        amount={amount}
        setAmount={(value) => setAmount(Number(value))}
      />
      {loadTried && !(amount > 0)
        && (
          <Alert AlertMsg="Amount must be greater than zero" />
        )}
      <Options
        defaultTitle="Select funding source"
        fundingSources={fundingSources}
        onChange={(option: FundingSource) => {
          setSelectedAccount(option);
        }}
        onEmptyOptionsSelected={() => {
          // navigate to add new account
        }}
      />
      {loadTried && !selectedAccount
        && (
          <Alert AlertMsg="Please select funding source" />
        )}
      <div className="action">
        <Button
          disabled={isLoading}
          className="zippy-btn"
          onClick={() => {
            loadFunds();
          }}
        >
          Load
        </Button>
      </div>
      <Modal show={cardExpired} onHide={() => setCardExpired(false)}>
        <Modal.Header closeButton>
          <Modal.Title>Error</Modal.Title>
        </Modal.Header>
        <Modal.Body>This Card is expired!</Modal.Body>
        <Modal.Footer>
          <Button
            className="zippy-btn"
            variant="secondary"
            onClick={() => setCardExpired(false)}
          >
            Remove Card
          </Button>
        </Modal.Footer>
      </Modal>
    </PageContainer>
  );
}
