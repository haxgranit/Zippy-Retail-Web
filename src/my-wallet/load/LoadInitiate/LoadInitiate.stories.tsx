import { ComponentStory, ComponentMeta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';
import React from 'react';
import LoadInitiate from './LoadInitiate';

export default {
  title: 'my-wallet/load/LoadInitiate',
  component: LoadInitiate,
} as ComponentMeta<typeof LoadInitiate>;

const Template: ComponentStory<typeof LoadInitiate> = () => (
  <BrowserRouter>
    <LoadInitiate />
  </BrowserRouter>
);

export const Default = Template.bind({});
Default.args = {};
Default.parameters = {};
