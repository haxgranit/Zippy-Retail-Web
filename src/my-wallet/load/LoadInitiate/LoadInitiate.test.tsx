import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import LoadInitiate from './LoadInitiate';

const ReactTestRenderer = require('react-test-renderer');

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useParams: () => ({
    stepId: 'details',
  }),
  useLocation: () => ({ state: { status: 'Success' } }),
  useNavigate: () => mockedUsedNavigate,
}));

describe('TransactionComplete Component', () => {
  it('matches the snapshot', () => {
    const tree = ReactTestRenderer.create(
      <BrowserRouter>
        <LoadInitiate />
      </BrowserRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
