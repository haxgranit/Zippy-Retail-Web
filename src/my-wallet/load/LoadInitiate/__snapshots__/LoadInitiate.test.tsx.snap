// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionComplete Component matches the snapshot 1`] = `
<div
  className=""
>
  <div
    className="zippy-cash-container-wrap"
  >
    <div
      className="backdrop"
    />
    <div
      className="zippy-cash-container"
    >
      <div
        className="zippy-cash-container-header"
      >
        <h1
          className="title"
        >
          Personal Banking
        </h1>
        <h2
          className="catch"
        >
          Made Fun With Zippy!
        </h2>
      </div>
      <div
        className="zippy-cash-container-body"
      >
        <div
          className="title"
        >
          Load wallet
        </div>
        <input
          className="amount form-control"
          defaultValue={0}
          onChange={[Function]}
          placeholder="0.00"
          step=".01"
          type="number"
        />
        <div
          className="position-relative"
        >
          <div
            className="dropbtn1"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            Select funding source
             
            <img
              alt=""
              src="https://dash.com.sg/assets/images/icons/arrow--down--grey.svg"
              width={30}
            />
          </div>
        </div>
        <div
          className="action"
        >
          <button
            className="zippy-btn btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Load
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
